src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,1048 @@
+ as [number, number, number] instead of GPUExtent3DDict.
+
+* copy_various_mip_levels: test that copying various mip levels works for all formats. Also covers special code paths:
+  - the physical size of the subresouce is not equal to the logical size
+  - bufferSize - offset < bytesPerImage * copyExtent.depth and copyExtent needs to be clamped
+
+* copy_with_no_image_or_slice_padding_and_undefined_values: test that when copying a single row we can set any bytesPerRow value and when copying a single\
+ slice we can set rowsPerImage to 0. Also test setting offset, rowsPerImage, mipLevel, origin, origin.{x,y,z} to undefined.
+
+* TODO: 
+  - add another initMethod which renders the texture
+  - because of expectContests 4-bytes alignment we don't test CopyT2B with buffer size not divisible by 4
+  - add tests for 1d / 3d textures
+`;
+
+import { params, poptions } from '../../../common/framework/params_builder.js';
+import { makeTestGroup } from '../../../common/framework/test_group.js';
+import { assert, unreachable } from '../../../common/framework/util/util.js';
+import {
+  kSizedTextureFormatInfo,
+  SizedTextureFormat,
+  kSizedTextureFormats,
+} from '../../capability_info.js';
+import { GPUTest } from '../../gpu_test.js';
+import { align } from '../../util/math.js';
+import { getTextureCopyLayout, TextureCopyLayout } from '../../util/texture/layout.js';
+
+interface TextureCopyViewWithRequiredOrigin {
+  texture: GPUTexture;
+  mipLevel: number | undefined;
+  origin: Required<GPUOrigin3DDict>;
+}
+
+/** Describes the function used to copy the initial data into the texture. */
+type InitMethod = 'WriteTexture' | 'CopyB2T';
+/** With PartialCopyT2B we do CopyT2B to check that the part of the texture we copied to with InitMethod
+ * matches the data we were copying and that we don't overwrite any data in the target buffer that we're not
+ * supposed to - that's primarily for testing CopyT2B functionality. With FullCopyT2B
+ * we do CopyT2B on the whole texture and check wether the part we copied to matches the data we were copying
+ * and that the nothing else was modified - that's primarily for testing WriteTexture and CopyB2T. */
+type CheckMethod = 'PartialCopyT2B' | 'FullCopyT2B';
+
+/** This describes in what form the arguments will be passed to WriteTexture/CopyB2T/CopyT2B.
+ * If undefined, then default values are passed as undefined instead of default values.
+ * If arrays, then GPUOrigin3D and GPUExtent3D are passed as [number, number, number]. */
+type ChangeBeforePass = 'none' | 'undefined' | 'arrays';
+
+/** Each combination of methods can assume that the ones before it were tested and work correctly. */
+const kMethodsToTest = [
+  /** We make sure that CopyT2B works when copying the whole texture for renderable formats:
+   * TODO */
+  /** Then we make sure that WriteTexture works for all formats: */
```

Comment:
fyi I don't think jsdoc-style comments have any effect on values/expressions like here, or on `g.test()` below. They just get treated as regular comments (which is fine though a little more verbose).